### PR TITLE
32 byte public keys in bip-taproot

### PR DESCRIPTION
**Pull request by jonasnick**
This is another attempt add 32 byte pubkeys to bip-schnorr, less ad-hoc than with a flag as in #51. The PR also contains a commit to make bip-taproot use this. The main difference to #51 is that this PR defines a separate signature scheme that's compatible with "short" pubkeys, both for signing and verification. This variant can be referred to as "bip-schnorr32" with the "verify32" and "sign32" algorithms.

Perhaps there are better arguments than what I mention in the commit for why the bip-schnorr32 variant does not replace the regular version.

The only place where bip-taproot switches to short pubkeys is output public keys. For script path spending it's necessary to reveal Y-bit in the control block.

This PR also fixes an existing inconsistency between bip-taproot and bip-schnorr, where in bip-taproot the point `Q` is given to the verification function as the public key, but bip-schnorr verification expects to get public key *bytes*.
A minor annoyance with short public keys in taproot outputs is that for `taproot_sign_script` you either need to remember to remember the Y-bit of the `output_public_key` or you need to recompute it.

(includes bugfix from #50)

### CODE REVIEW

**Comment by jonasnick** [comment]
> I think it will be good to specify a pubkey algorithm for both the regular scheme and the variant.

I added a pubkey algorithm earlier but then felt that was overkill because in `Signing32` we're saying 
> To sign ''m'' for public key ''bytes(dG)[1:33]'':

But either way is fine.

**Comment by real-or-random** [comment]
Oh yes, you're right. I've actually seen this. So yes, either way is fine. My feeling is that an explicit algorithm is slightly better because that sentence is easy to miss.

**Comment by sipa** [comment]
It doesn't seem necessary to reveal the P.y oddness in the control block. Instead, it could be made implicitly even. The Q.y oddness needs to always be revealed unless grinding/randomness is involved in creation.

Have you considered that option? I suspect it comes with additional complexities of output creation/signing, but I haven't thought through it all.

**Comment by real-or-random** [comment]
> It doesn't seem necessary to reveal the P.y oddness in the control block. Instead, it could be made implicitly even. The Q.y oddness needs to always be revealed unless grinding/randomness is involved in creation.

Did we ever consider the possibility of referring to P in the script? (Just something that came to my mind, not sure if that's too complex.)

**Comment by jonasnick** [comment]
> Did we ever consider the possibility of referring to P in the script?

Yes if I understand you correctly. It's part of @ajtown's anyprevout proposal (pubkey `0x01` refers to `P`)

File: bip-schnorr/reference.py [hunk 1]
```diff
@@ -88,14 +99,19 @@ def schnorr_verify(msg, pubkey, sig):
...
+        raise ValueError('The public key must be a 32-byte array.')
+    return schnorr_verify(msg, b'\x02' + pubkey, sig)
+
 #
 # The following code is only used to verify the test vectors.
 #
 import csv
 
-def test_vectors():
+def test_vectors(filename, schnorr_sign, schnorr_verify):
```
[the remark this answers is not in the document]
Comment (jonasnick) [reply bip-schnorr/reference.py]: fixed

File: bip-schnorr.mediawiki [hunk 2]
```diff
@@ -173,6 +188,34 @@ Many techniques are known for optimizing elliptic curve implementations. Several
...
+Input:
+* The secret key ''d' '': an integer in the range ''1..n-1''.
+* The message ''m'': a 32-byte array
+
+To sign ''m'' for public key ''bytes(dG)[1:33]'':
+* Let ''tag = bytes(dG)[0:1]''
+* Let ''d = -d' '' if ''tag &ne 0x02'', otherwise let ''d = d' ''.
+* Return signature obtained by running above regular signing algorithm with secret key ''d'' and message ''m''.
+
+Note that this specification is inefficient because the point ''dG'' is computed both here and in the regular signing algorithm.
```
[the remark this answers is not in the document]
Comment (jonasnick) [reply bip-schnorr.mediawiki]: fixed

File: bip-schnorr.mediawiki [hunk 3]
```diff
@@ -173,6 +188,34 @@ Many techniques are known for optimizing elliptic curve implementations. Several
...
+
+==== Verification32 ====
+
+Input:
+* The public key ''pk'': a 32-byte array
+* The message ''m'': a 32-byte array
+* A signature ''sig'': a 64-byte array
+
+The signature is valid if and only if the algorithm below does not fail.
+* Fail if above regular verification algorithm fails with public key ''0x02 || pk'', message ''m'' and signature ''sig''.
```
[the remark this answers is not in the document]
Comment (jonasnick) [reply bip-schnorr.mediawiki]: fixed

**Comment by jonasnick** [comment]
> I suggest we give both the schemes and all algorithms proper names, [...] The schemes could be bitcoin_schnorr and bitcoin_schnorr32, and then the algoritms would be bitcoin_schnorr_sign or bitcoin_schnorr32_verify etc. (feel free to use something else.)

I've been using `bip-schnorr` and `bip-schnorr32` for the schemes

File: bip-taproot.mediawiki [hunk 4]
```diff
@@ -62,22 +62,22 @@ In the text below, ''hash<sub>tag</sub>(m)'' is a shorthand for ''SHA256(SHA256(
...
... to every verifier, and thus only revealing its hash as an output would imply adding an additional 32 bytes to the witness. Furthermore, to maintain [https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-January/012198.html 128-bit collision security] for outputs, a 256-bit hash would be required anyway, which is comparable in size (and thus in cost for senders) to revealing the public key directly. While the usage of public key hashes is often said to protect against ECDLP breaks or quantum computers, this protection is very weak at best: transactions are not protected while being confirmed, and a very [https://twitter.com/pwuille/status/1108097835365339136 large portion] of the currency's supply is not under such protection regardless. Actual resistance to such systems can be introduced by relying on different cryptographic assumptions, but this proposal focuses on improvements that do not change the security model. Note that using P2SH-wrapped outputs only have 80-bit collision security. This is considered low, and is relevant whenever the output includes data from more than a single party (public keys, hashes, ...). </ref> If this is not a valid point on the curve, fail.
```
Comment (ajtowns) [inline bip-taproot.mediawiki]: "If this is not a valid point" -- "this" refers back to `q` now which is a 32-byte array; I think you need to ask if `point("0x02" || q)` is a valid point to make sense.

File: bip-taproot.mediawiki [hunk 5]
```diff
@@ -62,22 +62,22 @@ In the text below, ''hash<sub>tag</sub>(m)'' is a shorthand for ''SHA256(SHA256(
...
 ** The last stack element is called the control block ''c'', and must have length ''33 + 32m'', for a value of ''m'' that is an integer between 0 and 32, inclusive. Fail if it does not have such a length.
-** Let ''P = point(byte(2 + (c[0] & 1)) || c[1:33])''<ref>'''What is the purpose of the first byte of the control block?''' The first byte of the control block has three distinct functions:
+** Let ''P = point(byte(2 + (c[0] & 1)) || c[1:33])''<ref>'''What is the purpose of the first byte of the control block?''' The first byte of the control block has four distinct functions:
 * The low bit is used to denote the oddness of the Y coordinate of the ''P'' point.
+* The second bit is used to denote the oddness of the Y coordinate of the ''Q'' point.<ref>'''Why is the Y coordinate of the output public key required in a script path spend?''' Because ''Q'' is constructed by adding the taproot tweak to the internal public key ''P'', it cannot easily be ensured that ''Q'' has an even Y coordinate. Trying out multiple internal keys until there's such a ''Q'' is possible but undesirable and unnecessary since signalling the oddness of the Y coordinate only consumes an unused bit.</ref>
```
Comment (ajtowns) [inline bip-taproot.mediawiki]: I wonder if it wouldn't make for an easier explanation to treat the ''Q'' point as always even, and have the second bit used to denote whether the formula is negated?

File: bip-taproot.mediawiki [hunk 6]
```diff
@@ -62,22 +62,22 @@ In the text below, ''hash<sub>tag</sub>(m)'' is a shorthand for ''SHA256(SHA256(
...
 * By keeping the top two bits set to true, it can be guaranteed that scripts can be recognized without knowledge of the UTXO being spent, simplifying analysis. This is because such values cannot occur as first byte of the final stack element in either P2WPKH or P2WSH spends.
-* The remaining five bits are used for introducing new script versions that are not observable unless actually executed.
+* The remaining four bits are used for introducing new script versions that are not observable unless actually executed.
 </ref>. Fail if this point is not on the curve.
 ** Let ''l = c[0] & 0xfe'', the leaf version.
```
Comment (ajtowns) [inline bip-taproot.mediawiki]: `l = c[0] & 0xfc` if we're using the bottom two low bits, no? Would it make sense to say `l = ((c[0] & 0x3c) >> 2)` so the leaf version is just 0 to 15?

File: bip-taproot.mediawiki (hunk 5 above)
Comment (ajtowns) [reply bip-taproot.mediawiki]: (That would make the later test be `dQ = P + int(t)G` where `d=1` if `c[0]&2==0` and `d=-1` otherwise)

Comment (jonasnick) [reply bip-taproot.mediawiki]: I think so if we only have 32 byte keys

File: bip-taproot.mediawiki (hunk 4 above)
Comment (jonasnick) [reply bip-taproot.mediawiki]: Thanks. I thought I had that removed. It was wrong before and is even more wrong now (schnorr verification accepts a byte string and not a point)

File: bip-taproot.mediawiki (hunk 6 above)
Comment (jonasnick) [reply bip-taproot.mediawiki]: fixed

File: bip-schnorr/reference.py [hunk 7]
```diff
@@ -34,7 +34,7 @@ def bytes_from_point(P):
     return (b'\x03' if P[1] & 1 else b'\x02') + bytes_from_int(P[0])
 
 def point_from_bytes(b):
-    if b[0] in [b'\x02', b'\x03']:
+    if b[0:1] in [b'\x02', b'\x03']:
```
Comment (clarkmoody) [inline bip-schnorr/reference.py]: Doesn't `b[0:1]` return a list, such that you would be testing for `[x] in [y, z]` rather than `x in [y, z]`?

Comment (ajtowns) [reply bip-schnorr/reference.py]: depends what b is -- if it's a list, it'll return a smaller list, if it's a string it'll return a smaller string, if it's a bytes it'll return a smaller bytes.

Comment (clarkmoody) [reply bip-schnorr/reference.py]: Indeed. In the `list` case, this condition will always be `False`, but maybe we don't have to worry about that here, if only bytes are being thrown around.

Comment (jonasnick) [reply bip-schnorr/reference.py]: Doesn't seem like a problem given that this function is called `from_bytes`.

**Comment by jonasnick** [comment]
squashed

**Comment by sipa** [comment]
Superseded by #55.

**Pull request closed without merging**